File: src/main/java/uk/gov/justice/hmpps/prison/util/Extractors.java
package uk.gov.justice.hmpps.prison.util;

import java.math.BigDecimal;
import java.sql.Timestamp;
import java.time.LocalDate;
import java.util.Map;
import java.util.function.Function;

public class Extractors {
    public static Function<Object, String> extractString(final String field) {
        return m -> ((Map<String, String>) m).get(field);
    }

    public static Function<Object, Integer> extractInteger(final String field) {
        return m -> ((Map<String, BigDecimal>) m).get(field).intValue();
    }

    public static Function<Object, Long> extractLong(final String field) {
        return m -> ((Map<String, BigDecimal>) m).get(field).longValue();
    }

    public static Function<Object, LocalDate> extractDate(final String field) {
        return m -> ((Map<String, Timestamp>) m).get(field).toLocalDateTime().toLocalDate();
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/util/Extractors.kt
@file:Suppress("UNCHECKED_CAST")

package uk.gov.justice.hmpps.prison.util

import java.math.BigDecimal
import java.sql.Timestamp
import java.time.LocalDate
import java.util.function.Function

object Extractors {
  @JvmStatic
  fun extractString(field: String?): Function<Any, String?> = Function { m: Any ->
    (m as Map<String?, String?>)[field]
  }

  @JvmStatic
  fun extractInteger(field: String?): Function<Any, Int> = Function { m: Any ->
    (m as Map<String?, BigDecimal>)[field]!!.toInt()
  }

  @JvmStatic
  fun extractLong(field: String?): Function<Any, Long> = Function { m: Any ->
    (m as Map<String?, BigDecimal>)[field]!!.toLong()
  }

  @JvmStatic
  fun extractDate(field: String?): Function<Any, LocalDate> = Function { m: Any ->
    (m as Map<String?, Timestamp>)[field]!!.toLocalDateTime().toLocalDate()
  }
}
